File: chapter_3_stacks_and_queues/js/3.3-stackOfPlates.js
const { Stack } = require("./stack")

class StackWithCapacity extends Stack {
  constructor(capacity) {
    super()
    this.capacity = capacity
  }

  push(value) {
    if (this.isFull()) throw new Error("stack is full")
    this.stack.push(value)
    this.length++
  }

  isFull() {
    return this.length === this.capacity
  }
}

swc = new StackWithCapacity(5)
swc.push(1)
swc.push(1)
swc.push(1)
swc.push(1)
swc.push(1)
// swc.push(1) // error
console.log(swc)
/*
StackWithCapacity { stack: [ 1, 1, 1, 1, 1 ], length: 5, capacity: 5 }
*/

class SetOfStacks {
  constructor(stackCapacity) {
    this.stacks = []
    this.stacksLength = 0
    this.stackCapacity = stackCapacity
  }

  push(value) {
    const last = this.getLastStack()
    if (last != null && !last.isFull()) {
      /* add to last stack */
      last.push(value)
    } else {
      /* create new stack */
      const newStack = new StackWithCapacity(this.stackCapacity)
      newStack.push(value)
      this.stacks.push(newStack)
      this.stacksLength++
    }
  }

  pop() {
    const last = this.getLastStack()
    if (last === null) throw new Error("empty stack of stacks")
    const value = last.pop()
    /* If stack is empty, remove it from stacks */
    if (last.length === 0) {
      this.stacks.pop()
      this.stacksLength--
    }
    return value
  }

  getLastStack() {
    if (this.stacksLength === 0) return null
    return this.stacks[this.stacksLength - 1]
  }
}

const sos = new SetOfStacks(3)
console.log(sos)
/*
SetOfStacks { stacks: [], stacksLength: 0, stackCapacity: 3 }
*/
const arr = [1, 2, 3, 4, 5, 6, 7]
for (let i of arr) sos.push(i)
console.log(sos)
/*
SetOfStacks {
  stacks:
   [ StackWithCapacity { stack: [Array], length: 3, capacity: 3 },
     StackWithCapacity { stack: [Array], length: 3, capacity: 3 },
     StackWithCapacity { stack: [Array], length: 3, capacity: 3 },
     StackWithCapacity { stack: [Array], length: 1, capacity: 3 } ],
  stacksLength: 4,
  stackCapacity: 3 }
*/
/* remove 4 items */
for (let i of Array(4)) sos.pop()
console.log(sos)
/*
SetOfStacks {
  stacks:
   [ StackWithCapacity { stack: [Array], length: 3, capacity: 3 } ],
  stacksLength: 1,
  stackCapacity: 3 }
*/

File: chapter_3_stacks_and_queues/python/3.3-stack_of_plates.py
Stack = __import__('stack').Stack


class StackWithCapacity(Stack):
    def __init__(self, capacity=5):
        super().__init__()
        self.capacity = capacity

    def push(self, value):
        if self.is_full():
            raise Exception('stack is full')
        super().push(value)

    def is_full(self):
        return self.length == self.capacity


class SetOfStacks:
    def __init__(self):
        self.stacks = []
        self.length = 0

    def push(self, value):
        last = self.get_last_stack()
        if last == None or last.is_full():
            # create new stack
            new_stack = StackWithCapacity()
            new_stack.push(value)
            self.stacks.append(new_stack)
            self.length += 1
        else:
            # add to last stack
            last.push(value)

    def pop(self):
        last = self.get_last_stack()
        if last == None:
            raise Exception('stacks are empty')
        value = last.pop()
        if last.is_empty():
            # If stack is empty, remove it from stacks
            self.stacks.pop()
            self.length -= 1
        return value

    def get_last_stack(self):
        if self.length == 0:
            return None
        return self.stacks[self.length - 1]


if __name__ == "main":
    swc = StackWithCapacity(5)
    for i in range(5):
        swc.push(i)
    print(swc.__dict__)
    # {'stack': [0, 1, 2, 3, 4], 'length': 5, 'capacity': 5}': 5, 'capacity': 5}

    sos = SetOfStacks()
    for i in range(11):
        sos.push(i)
    print([x.stack for x in sos.stacks])
    # [[0, 1, 2, 3, 4], [5, 6, 7, 8, 9], [10]] [10]]
    print(sos.pop())  # 10
    print(sos.pop())  # 9
    print([x.stack for x in sos.stacks])
    # [[0, 1, 2, 3, 4], [5, 6, 7, 8]]
